src/main/java/org/truffleruby/core/array/ArrayAppendOneNode.java

```diff
@@ -60,9 +60,9 @@ public DynamicObject appendOneSameType(DynamicObject array, Object value,
     // Append forcing a generalization
 
     @Specialization(guards = {
-            "strategy.matches(array)", "!strategy.accepts(value)", "valueStrategy.specializesFor(value)",
+            "strategy.matches(array)", "valueStrategy.specializesFor(value)",
```

Comment:
That's a tricky change to review. But it seems right.
The problem would be if we take this specialization if strategy and valueStrategy are the same (and mutable).
But this should not be the case since specializations are tried in order and the specialization should be instantiated in such a case.
Could you maybe just add an assert in the body that strategy != valueStrategy?